File: examples/spanner-client/php/analyze_and_render.php
#!/usr/bin/env php
<?php

declare(strict_types=1);

require __DIR__ . '/vendor/autoload.php';

use Google\Cloud\Spanner\V1\Client\SpannerClient;
use Google\Cloud\Spanner\V1\CreateSessionRequest;
use Google\Cloud\Spanner\V1\DeleteSessionRequest;
use Google\Cloud\Spanner\V1\ExecuteSqlRequest;
use Google\Cloud\Spanner\V1\TransactionOptions;
use Google\Cloud\Spanner\V1\TransactionOptions\PBReadOnly;
use Google\Cloud\Spanner\V1\TransactionSelector;

function fetchQueryPlan(CliOptions $opts): Google\Cloud\Spanner\V1\QueryPlan
{
    $databasePath = sprintf(
        'projects/%s/instances/%s/databases/%s',
        $opts->project,
        $opts->instance,
        $opts->database,
    );

    $client = new SpannerClient();
    $session = $client->createSession((new CreateSessionRequest())->setDatabase($databasePath));

    try {
        $request = (new ExecuteSqlRequest())
            ->setSession($session->getName())
            ->setSql($opts->sql)
            ->setQueryMode(QueryModeOption::spannerQueryMode($opts->queryMode))
            ->setTransaction(
                (new TransactionSelector())->setSingleUse(
                    (new TransactionOptions())->setReadOnly(new PBReadOnly()),
                ),
            );

        $result = $client->executeSql($request);
        if ($opts->queryMode === 'PROFILE') {
            foreach ($result->getRows() as $_row) {
            }
        }

        $plan = $result->getStats()?->getQueryPlan();
        if ($plan === null) {
            throw new RuntimeException('QueryPlan missing from ResultSetStats');
        }

        return $plan;
    } finally {
        $client->deleteSession((new DeleteSessionRequest())->setName($session->getName()));
    }
}

try {
    $opts = CliOptions::parse();
    $plan = fetchQueryPlan($opts);
    fwrite(
        STDOUT,
        SpannerAdapter::renderQueryPlan($plan, QueryModeOption::renderMode($opts->queryMode)),
    );
    exit(0);
} catch (InvalidArgumentException $e) {
    fwrite(STDERR, $e->getMessage() . PHP_EOL);
    exit(2);
} catch (Throwable $e) {
    fwrite(STDERR, $e->getMessage() . PHP_EOL);
    exit(1);
}

File: examples/spanner-client/ruby/analyze_and_render.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'google/cloud/spanner/v1'

require_relative 'cli_options'
require_relative 'query_mode_option'
require_relative 'spanner_adapter'

def fetch_query_plan(opts)
  database_path = "projects/#{opts.project}/instances/#{opts.instance}/databases/#{opts.database}"
  client = Google::Cloud::Spanner::V1::Spanner::Client.new

  session = client.create_session(
    Google::Cloud::Spanner::V1::CreateSessionRequest.new(database: database_path)
  )
  begin
    result = client.execute_sql(
      Google::Cloud::Spanner::V1::ExecuteSqlRequest.new(
        session: session.name,
        sql: opts.sql,
        transaction: { single_use: { read_only: {} } },
        query_mode: QueryModeOption.spanner_query_mode(opts.query_mode)
      )
    )

    result.rows.each { |_| } if opts.query_mode == 'PROFILE'

    stats = result.stats
    raise 'QueryPlan missing from ResultSetStats' if stats.nil? || stats.query_plan.nil?

    stats.query_plan
  ensure
    client.delete_session(
      Google::Cloud::Spanner::V1::DeleteSessionRequest.new(name: session.name)
    )
  end
end

def main
  opts = CliOptionsParser.parse
  plan = fetch_query_plan(opts)
  print SpannerAdapter.render_query_plan(plan, mode: QueryModeOption.render_mode(opts.query_mode))
  0
rescue ArgumentError => e
  warn e.message
  2
rescue StandardError => e
  warn e.message
  1
end

exit main if $PROGRAM_NAME == __FILE__
